src/decoders/gradebook.ts

```
@@ -0,0 +1,48 @@
+import { GradeBook, GradeBookSubject } from "~/models/gradebook";
+import { decodeSubject } from "./subject";
+import { gradebookPDF } from "~/api";
+import { Period, SessionHandle } from "~/models";
+
+/**
+ * @param gradeBookData response from PageBulletins
+ */
+export const decodeGradeBook = async (session: SessionHandle, period: Period, gradeBookData: any): Promise<GradeBook> => {
+  // When bad period is used, the return is `{ data: {}, nom: 'PageBulletins' }` but the session don't expire.
+  if (Object.keys(gradeBookData).length == 0 || gradeBookData.message)
+    return { available: false, message: gradeBookData.message ?? undefined };
```

Would be better if you throw an error right there instead of returning a different object.

Library user could handle it using a try/catch.